### PR TITLE
Handle avatar image error

### LINKED ISSUE #402: Avatars sometime show broken images
We should make sure that image links in avatars are valid.

### PR DESCRIPTION
**Pull request by MidhunSureshR**
Fixes #402 
Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

### CODE REVIEW

File: src/platform/web/ui/avatar.js [hunk 1]
```diff
@@ -65,6 +70,8 @@ export class AvatarView extends BaseUpdateView {
         this._avatarLetterChanged();
         this._avatarTitleChanged();
         this._root = renderStaticAvatar(this.value, this._size);
+        const image = this._root.firstChild;
+        image?.addEventListener("error", () => this.setAvatarError());
```
Comment (bwindels) [inline src/platform/web/ui/avatar.js]: Perhaps also listen for load here and in both handlers, remove the listeners of both, so once an image is loaded or failed to load, there are no listeners anymore.

File: src/platform/web/ui/avatar.js [hunk 2]
```diff
@@ -65,6 +70,8 @@ export class AvatarView extends BaseUpdateView {
         this._avatarLetterChanged();
         this._avatarTitleChanged();
         this._root = renderStaticAvatar(this.value, this._size);
+        const image = this._root.firstChild;
```
Comment (bwindels) [inline src/platform/web/ui/avatar.js]: Can we also do this for static avatars (e.g. done with renderStaticAvatar and not binding to the view model)? We'll need to handle both separately though I think, as we need to know in update if the avatar failed to load to add or remove the background color class.

**Review by bwindels — COMMENTED** [review]
Thanks! Looks mainly good, just a few small nits.

File: src/platform/web/ui/avatar.js [hunk 3]
```diff
@@ -57,6 +57,20 @@ export class AvatarView extends BaseUpdateView {
...
+        const handleAvatarError = (e) => {
+            const image = e.target;
+            image.removeEventListener("load", removeErrorHandler);
+            this._setAvatarError();
+        };
+        const removeErrorHandler = (e) => {
+            const image = e.target;
+            image.removeEventListener("error", handleAvatarError);
+        };
+        image?.addEventListener("error", handleAvatarError);
```
Comment (bwindels) [inline src/platform/web/ui/avatar.js]: don't you need to remove both handlers on either event? I think you can use the same handler for both events even, and set `this._avatarError` to `e.type === "error";`.

Comment (MidhunSureshR) [reply src/platform/web/ui/avatar.js]: Oops! Sorry about that

File: src/platform/web/Platform.js [hunk 4]
```diff
@@ -189,6 +190,7 @@ export class Platform {
                 this._disposables.track(disposable);
             }
         }
+        this._container.addEventListener("error", handleAvatarError, true);
```
Comment (bwindels) [inline src/platform/web/Platform.js:193]: Can you add a closure to disposables to remove this again when dispose gets called? `adaptUIOnVisualViewportResize` has an example of this.

File: src/platform/web/ui/avatar.js [hunk 5]
```diff
@@ -108,16 +25,29 @@ export function renderStaticAvatar(vm, size, extraClasses = undefined) {
...
         [`size-${size}`]: true,
-        [`usercolor${vm.avatarColorNumber}`]: !hasAvatar,
+        [`usercolor${vm.avatarColorNumber}`]: true
     });
     if (extraClasses) {
         avatarClasses += ` ${extraClasses}`;
     }
     const avatarContent = hasAvatar ? renderImg(vm, size) : text(vm.avatarLetter);
-    return tag.div({className: avatarClasses}, [avatarContent]);
+    return tag.div({className: avatarClasses, 'data-avatar-letter': vm.avatarLetter}, [avatarContent]);
```
Comment (bwindels) [inline src/platform/web/ui/avatar.js]: You could only set the letter if `hasAvatar` is true

File: src/platform/web/ui/avatar.js [hunk 6]
```diff
@@ -108,16 +25,29 @@ export function renderStaticAvatar(vm, size, extraClasses = undefined) {
...
+    const element = e.target;
+    const parent = element.parentElement;
+    return element.tagName === "IMG" && parent.classList.contains("avatar");
+}
+
+export function handleAvatarError(e) {
+    if (!isAvatarEvent(e)) { return; }
+    const parent = e.target.parentElement;
+    const avatarLetter = parent.getAttribute("data-avatar-letter");
+    parent.textContent = avatarLetter;
```
Comment (bwindels) [inline src/platform/web/ui/avatar.js:59]: Wouldn't you also need to hide the img somehow? (and unhide if the binding in AvatarView changes the `src` property)

File: src/platform/web/ui/AvatarView.js [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    update(vm) {
+        // important to always call _...changed for every prop 
+        if (this._avatarUrlChanged()) {
+            // avatarColorNumber won't change, it's based on room/user id
+            const bgColorClass = `usercolor${vm.avatarColorNumber}`;
+            if (vm.avatarUrl(this._size)) {
+                this._root.replaceChild(renderImg(vm, this._size), this._root.firstChild);
+                this._root.classList.remove(bgColorClass);
```
Comment (bwindels) [inline src/platform/web/ui/AvatarView.js:69]: The background color class should not be removed or set here I imagine, but just always left as it was by the initial render in `renderStaticAvatar`. You would however need to add/remove a class to show or hide the image when it fails to load I think?

**Review by bwindels — COMMENTED** [review]
Looking good, just a few more things.

File: src/platform/web/ui/avatar.js (hunk 6 above)
Comment (bwindels) [reply src/platform/web/ui/avatar.js:59]: textContent already clears the image

File: src/platform/web/ui/AvatarView.js [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+                this._root.classList.add(bgColorClass);
+            }
+        }
+        const hasAvatar = !!vm.avatarUrl(this._size);
+        if (this._avatarTitleChanged() && hasAvatar) {
+            const img = this._root.firstChild;
+            img.setAttribute("title", vm.avatarTitle);
+        }
+        if (this._avatarLetterChanged() && !hasAvatar) {
+            this._root.firstChild.textContent = vm.avatarLetter;
```
Comment (bwindels) [inline src/platform/web/ui/AvatarView.js]: Can you do `this._root.textContent = ...` here as well like you did in handleAvatarError?

File: src/platform/web/ui/AvatarView.js [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+                this._root.replaceChild(renderImg(vm, this._size), this._root.firstChild);
+                this._root.classList.remove(bgColorClass);
+            } else {
+                this._root.replaceChild(text(vm.avatarLetter), this._root.firstChild);
+                this._root.classList.add(bgColorClass);
+            }
+        }
+        const hasAvatar = !!vm.avatarUrl(this._size);
+        if (this._avatarTitleChanged() && hasAvatar) {
+            const img = this._root.firstChild;
```
Comment (bwindels) [inline src/platform/web/ui/AvatarView.js]: Can you check here now if this is indeed an image? As we might have removed it on error ...

File: src/platform/web/ui/AvatarView.js [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+    update(vm) {
+        // important to always call _...changed for every prop 
+        if (this._avatarUrlChanged()) {
+            // avatarColorNumber won't change, it's based on room/user id
+            const bgColorClass = `usercolor${vm.avatarColorNumber}`;
+            if (vm.avatarUrl(this._size)) {
+                this._root.replaceChild(renderImg(vm, this._size), this._root.firstChild);
+                this._root.classList.remove(bgColorClass);
+            } else {
+                this._root.replaceChild(text(vm.avatarLetter), this._root.firstChild);
```
Comment (bwindels) [inline src/platform/web/ui/AvatarView.js]: Can you do `this._root.textContent = ...` here as well like you did in handleAvatarError?

File: src/platform/web/ui/avatar.js [hunk 11]
```diff
@@ -108,16 +25,36 @@ export function renderStaticAvatar(vm, size, extraClasses = undefined) {
...
+    const parent = element.parentElement;
+    return element.tagName === "IMG" && parent.classList.contains("avatar");
+}
+
+export function handleAvatarError(e) {
+    if (!isAvatarEvent(e)) { return; }
+    const parent = e.target.parentElement;
+    const avatarColorNumber = parent.getAttribute("data-avatar-color");
+    parent.classList.add(`usercolor${avatarColorNumber}`);
+    const avatarLetter = parent.getAttribute("data-avatar-letter");
```
Comment (bwindels) [inline src/platform/web/ui/avatar.js:58]: actually, looks like you want to use innerText to make the  go away in IE11
https://developer.mozilla.org/en-US/docs/Web/API/Node/textContent#differences_from_innertext

**Review by bwindels — APPROVED** [review]
Looks good, thanks! Did you test that it works as well?

**Comment by MidhunSureshR** [comment]
buggy
Before
memberlist
Works with memberlist :heavy_check_mark: 
timeline
Works with timeline :heavy_check_mark: 
transparent
Adds white background by default :heavy_check_mark:
Is that good enough :thinking: or do we need to test with a dummy view with invalid avatars?

**Comment by bwindels** [comment]
That's good enough for testing for me! Let's merge it!

**Pull request merged**